src/java.net.http/share/classes/jdk/internal/net/http/PullPublisher.java

```diff
@@ -25,62 +25,71 @@
 
 package jdk.internal.net.http;
 
-import java.util.Iterator;
 import java.util.concurrent.Flow;
+
 import jdk.internal.net.http.common.Demand;
 import jdk.internal.net.http.common.SequentialScheduler;
 
 /**
- * A Publisher that publishes items obtained from the given Iterable. Each new
- * subscription gets a new Iterator.
+ * A {@linkplain Flow.Publisher publisher} that publishes items obtained from the given {@link CheckedIterable}.
+ * Each new subscription gets a new {@link CheckedIterator}.
  */
 class PullPublisher<T> implements Flow.Publisher<T> {
 
-    // Only one of `iterable` and `throwable` can be non-null. throwable is
+    // Only one of `iterable` or `throwable` should be null, and the other non-null. throwable is
```

Comment:
`PullPublisher` logic is branching on the `throwable == null` condition on several places – this increases code size and complexity. Plus, all `PullPublisher::new` call sites know at compile time whether `throwable` is null or not. I think `PullPublisher` should only accept a `CheckedIterable`, and we should create a new, say, `ImmediatelyFailingPublisher` to cover the cases where a `Throwable` needs to be passed to subscribers verbatim. I did not carry out this improvement in this PR to avoid scope creep, but I'm inclined to create a ticket for it. WDYT?